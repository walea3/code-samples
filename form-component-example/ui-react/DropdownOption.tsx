import cx from 'classnames';
import React from 'react';

import { getDataRef } from '@nel-ui/js-helpers';
import { Text } from '@nel-ui/typography-react';

import FormStyles from './FormStyles';

export interface DropdownSelectableOption {
  id: string;
  value: string | number;
  text: React.ReactText;
}

export interface DropdownOptionProps {
  id: DropdownSelectableOption['id'];
  value: DropdownSelectableOption['value'];
  children: DropdownSelectableOption['text'];
  onChange: (value: DropdownOptionProps['value']) => void;
  selected?: boolean;
  active?: boolean;
  'data-ref'?: string;
}

export const DropdownOption = ({
  onChange,
  children,
  value,
  active = false,
  selected = false,
  'data-ref': dataRef,
  ...props
}: DropdownOptionProps) => (
  <FormStyles>
    {({ classes }) => (
      <Text
        as="li"
        data-ref={getDataRef(dataRef, 'dropdownOption')}
        onMouseDown={() => onChange(value)}
        className={cx(classes.dropdownOption, {
          [classes.selectedDropdownOption]: selected,
          [classes.activeDropdownOption]: active,
        })}
        role="option"
        aria-selected={active}
        {...props}
      >
        {children}
      </Text>
    )}
  </FormStyles>
);

DropdownOption.displayName = 'Dropdown.Option';
