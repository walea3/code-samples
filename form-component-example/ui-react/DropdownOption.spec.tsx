import { mountWithNEL, renderWithNEL } from '@nel-ui/tooling-react/test-utils';
import React from 'react';

import { DropdownOption } from './DropdownOption';

describe('<DropdownOption />', () => {
  const defaultProps = {
    id: 'option-1',
    value: 'option value',
    children: 'Dropdown option',
  };

  describe('@renders', () => {
    test.each`
      name                  | props
      ${'default'}          | ${{}}
      ${'when selected'}    | ${{ selected: true }}
      ${'when active'}      | ${{ active: true }}
      ${'with data-ref'}    | ${{ 'data-ref': 'uid' }}
      ${'with other props'} | ${{ id: 'test' }}
    `('$name', ({ props }) =>
      expect(
        renderWithNEL(<DropdownOption {...defaultProps} {...props} />)
      ).toMatchSnapshot()
    );
  });

  describe('@events', () => {
    const onChange = jest.fn();

    test('on mouse down calls on change handler with option value', () => {
      const { root, wrapper } = mountWithNEL(
        <DropdownOption {...defaultProps} {...{ onChange }} />
      );

      wrapper
        .find('[role="option"]')
        .first()
        .simulate('mouseDown');

      expect(onChange).toHaveBeenCalledTimes(1);
      expect(onChange).toHaveBeenCalledWith(defaultProps.value);

      root.unmount();
    });
  });
});
