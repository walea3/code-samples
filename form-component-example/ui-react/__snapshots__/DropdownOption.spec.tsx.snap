// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownOption /> @renders default 1`] = `
<li
  aria-selected="false"
  class="nel-dropdownOption nel-text"
  data-ref="dropdownOption.text"
  id="option-1"
  role="option"
>
  Dropdown option
</li>
`;

exports[`<DropdownOption /> @renders when active 1`] = `
<li
  aria-selected="true"
  class="nel-dropdownOption nel-activeDropdownOption nel-text"
  data-ref="dropdownOption.text"
  id="option-1"
  role="option"
>
  Dropdown option
</li>
`;

exports[`<DropdownOption /> @renders when selected 1`] = `
<li
  aria-selected="false"
  class="nel-dropdownOption nel-selectedDropdownOption nel-text"
  data-ref="dropdownOption.text"
  id="option-1"
  role="option"
>
  Dropdown option
</li>
`;

exports[`<DropdownOption /> @renders with data-ref 1`] = `
<li
  aria-selected="false"
  class="nel-dropdownOption nel-text"
  data-ref="uid.dropdownOption.text"
  id="option-1"
  role="option"
>
  Dropdown option
</li>
`;

exports[`<DropdownOption /> @renders with other props 1`] = `
<li
  aria-selected="false"
  class="nel-dropdownOption nel-text"
  data-ref="dropdownOption.text"
  id="test"
  role="option"
>
  Dropdown option
</li>
`;
